### PR TITLE
improve spawn in Android apps

### PR DESCRIPTION
**Pull request by muhzii**
allow using the activity param in case the app doesn't have a launcher activity

### CODE REVIEW

File: src/linux/agent/system-server.js [hunk 1]
```diff
@@ -85,12 +86,35 @@ rpc.exports = {
   },
   startActivity: function (pkg, activity) {
     return performOnJavaVM(function () {
-      var intent = packageManager.getLaunchIntentForPackage(pkg);
-      if (intent === null)
+      var appInstalled = false;
+      var apps = packageManager.getInstalledApplications(GET_META_DATA);
+      for (var i = 0; i < apps.size(); i++) {
```
Comment (oleavr) [inline src/linux/agent/system-server.js]: Let's only read the `size` once, by moving it above the loop:

```suggestion
      var numApps = apps.size();
      for (var i = 0; i !== numApps; i++) {
```

(And also use the same loop idiom as elsewhere for consistency.)

File: src/linux/agent/system-server.js [hunk 2]
```diff
@@ -85,12 +86,35 @@ rpc.exports = {
...
+          break;
+        }
+      }
+      if (!appInstalled)
         throw new Error("Unable to find application with identifier '" + pkg + "'");
 
-      if (activity !== null)
+      var intent = packageManager.getLaunchIntentForPackage(pkg);
+      if (intent === null && activity === null)
+        throw new Error("Unable to find a front-door activity");
```
Comment (oleavr) [inline src/linux/agent/system-server.js]: Strings should be this style:

```suggestion
        throw new Error('Unable to find a front-door activity');
```

(Though we've made a few exceptions in this file for cases where we want to use single-quotes in the string literal.)

File: src/linux/agent/system-server.js [hunk 3]
```diff
@@ -85,12 +86,35 @@ rpc.exports = {
...
+
+      if (intent === null) {
+        intent = Intent.$new();
+        intent.setAction("android.intent.action.MAIN");
```
Comment (oleavr) [inline src/linux/agent/system-server.js]:
```suggestion
        intent.setAction('android.intent.action.MAIN');
```

File: src/linux/agent/system-server.js [hunk 4]
```diff
@@ -85,12 +86,35 @@ rpc.exports = {
...
+      }
+
+      if (activity !== null) {
+        var pkgInfo = packageManager.getPackageInfo(pkg, GET_ACTIVITIES);
+        var activities = pkgInfo.activities.value.map((activityInfo) => activityInfo.name.value);
```
Comment (oleavr) [inline src/linux/agent/system-server.js]:
```suggestion
        var activities = pkgInfo.activities.value.map((activityInfo) => activityInfo.name.value);
```

Should stick to ES5 in the internal agents so they can also be used with Duktape. (Even if we currently use V8 for this particular agent.)

File: src/linux/agent/system-server.js [hunk 5]
```diff
@@ -85,12 +86,35 @@ rpc.exports = {
...
+        if (activities.indexOf(activity) < 0)
```
Comment (oleavr) [inline src/linux/agent/system-server.js]:
```suggestion
        if (activities.indexOf(activity) === -1)
```

(For consistency.)

File: src/linux/agent/system-server.js [hunk 6]
```diff
@@ -100,17 +101,20 @@ rpc.exports = {
...
       }
 
       if (activity !== null) {
+        var activities = [];
         var pkgInfo = packageManager.getPackageInfo(pkg, GET_ACTIVITIES);
-        var activities = pkgInfo.activities.value.map((activityInfo) => activityInfo.name.value);
-        if (activities.indexOf(activity) < 0)
+        pkgInfo.activities.value.forEach(function (activityInfo) {
+          activities.push(activityInfo.name.value);
+        });
```
Comment (oleavr) [inline src/linux/agent/system-server.js]: Sorry, I wasn't clear – `map()` is available, only the arrow function part needed changing.

Comment (muhzii) [reply src/linux/agent/system-server.js]: of course

**Pull request merged**